### PR TITLE
[ConstraintSystem] Make relabeling logic consistent

### PR DESCRIPTION
**Pull request by omochi**
# Summary

It introduce `CallArgumentMatchResult` and `RelabelingMapping` type.
Using them make behavior of `matchCallArgument`, calculating impact of relabeling fix and relabeling diagnostic consistent.

@xedin please review this.

# Motivation

`matchCallArgument`, `ArgumentFailureTracker::relabelArguments`, `diagnoseArgumentLabelError` have similar duplicated logics for each purpose.

Current such design makes improvement this are difficult.

Some information at `matchCallArgument` is not passed to other stages.

So some problem is happen in corner case.

I also heard this idea here.

https://forums.swift.org/t/question-about-overlapped-multiple-fixits-from-diagnostics/34488/2

It's good if I could follow the intention what @xedin tell me `TODO` here.

## Diagnostics of relabeling does not consider default parameter.

```swift
func f(aa: Int, bb: Int, cc: Int = 0) {}
f(bbx: 0)
// incorrect argument labels in call (have 'bbx:', expected 'aa:bb:cc:')
```

Expecting of label `cc:` isn't necessary.

## Diagnostics of relabeling does not consider variadics parameter.

```swift
func f(aa: Int..., bb: Int, cc: Int) {}
f(aa: 1, 2, 3, bbx: 4, cc: 5)
// incorrect argument label in call (have 'aa:_:_:bbx:cc:', expected 'aa:_:_:bb:cc:')
```

Each entry of variadics change into unlabeled individual arguments on label.

## Calculating impact factor of relabeling fix does not consider parameter claiming.

```swift
func f(aa: Int, bb: Int) {}
f(aa: 1, aa: 2)
// incorrect argument label in call (have 'aa:aa:', expected 'aa:bb:')
// Score: 2 0 0 0 0 0 0 0 0 0 0 0
```

Score is `1(Relabeling) + 1(OoO)`.
But second `aa` argument shouldn't be considered as out of order from parameter `aa:`.
Because parameter `aa:` is claimed by first `aa` argument.

---

I think these problems are coming from these duplicated logics.

# Design

To resolve these issue fundamentally,
It introduce two types which is `CallArgumentMatchResult` and `RelabelingMapping`.

They are both just value type holding information.

It split `matchCallArgument` into two overloads.

First one build `CallArgumentMatchResult` and have no side effect.
So it doesn't take `MatchCallArgumentListener`.
Second one keeps compatibility.
It call first one and use result.
It invoke `MatchCallArgumentListener` based on result.

I think that this refactoring makes it more readable in particular about tracking side effect and branching flow.

`CallArgumentMatchResult::buildRelabelingMapping` build `RelabelingMapping`.

`RelabelingMapping` holds mapping information between arguments and parameters, judged label mismatch category.

It changes `RelabelArguments` and `LabelingFailure` to take `RelabelingMapping` instead of `ArrayRef<Identifier>`.

It split `diagnoseArgumentLabelError` into two overloads.
First one takes `RelabelingMapping` and make diagnostics and fix-its from it.
Second one keeps compatibility.
It build `RelabelingMapping` using original logic and calls first one.

`LabelingFailure` calls first one of `diagnoseArgumentLabelError` with own `RelabelingMapping`.

Because `RelabelingMapping` is just a value and built from `CallArgumentMatchResult`,
they can remove duplicated logics and use shared consistent result.

# Achievement and change of test result

Diff of `ArgumentFailureTracker::relabelArguments` and `diagnoseArgumentLabelError` seems to be good for me.
Each original logics of label iteration and matching are removed.
They are just only convert passed argument into their purpose now.

##  label of variadics

```
-variadics1(1, 2, 3, 4, 5, x: 6, y: 7) // expected-error {{incorrect argument labels in call (have '_:_:_:_:_:x:y:', expected 'x:y:_:')}} {{12-12=x: }} {{15-15=y: }} {{27-30=}} {{33-36=}}
+variadics1(1, 2, 3, 4, 5, x: 6, y: 7) // expected-error {{incorrect argument labels in call (have '_:x:y:', expected 'x:y:_:')}} {{12-12=x: }} {{27-28=y}} {{33-36=}}
```

Diagnose and fix-it now knows variadics.
After fix it, code is going to be `variadics1(x: 1, 2, 3, 4, 5, y: 6, 7)`.

Following is same.

```
-    f(alpha: 0, 1, bravox: 3) // expected-error {{incorrect argument label in call (have 'alpha:_:bravox:', expected 'alpha:_:bravo:')}}
+    f(alpha: 0, 1, bravox: 3) // expected-error {{incorrect argument label in call (have 'alpha:bravox:', expected 'alpha:bravo:')}}
```

## default parameter

```
func f1(_ a: Bool = false, _ b: Int = 0, c: String = "", _ d: [Int] = []) {}

-    f1(b: "2", [3]) // expected-error {{incorrect argument labels in call (have 'b:_:', expected '_:_:c:_:')}}
+    f1(b: "2", [3]) // expected-error {{incorrect argument labels in call (have 'b:_:', expected '_:c:')}}
```

Diagnose and fix-it knows that `_ a:` and `_ d:` aren't needed as default value.

# Trouble at coding

I am in some trouble at coding.

## `getArgumentListExprFor`

`RelabelArguments` also created at `matchFunctionTypes`.
To keep same behavior, I need helper function `FailureDiagnostics::getArgumentListExprFor`.
So I publish this at `constraints::getArgumentListExprFor`.
I worried about this design.

## `#include`

I put `RelabelingArgument` at `CSFix.h`.
So I need `#include "ConstraintSystem.h"` at `MiscDiagnostics.cpp`.
First time I wrote `#include "CSFix.h`.
But after merged this,
https://github.com/apple/swift/pull/30411
including order was twisted and I changed.

I moved original mapping logic of `diagnoseArgumentLabelError` to `RelabelingMapping::build`.
It calls `getOriginalArgumentList`.
So I need `#include "swift/Sema/IDETypeChecking.h"` at `CSFix.cpp`.

I feel these `#include` addition is not good.
Please advice me.

# Future direction

If this PR is acceptable, I will more refactor on changes.

## Refactor `matchCallArgument`

I currently prioritized compatibility and ease of reviewing.
So code has duplicated variables to track intermediate state.
For example, local variable `parameterBindings` could be removed as using `result`.

## Improve data structure of `CallArgumentMatchResult`

It can be changed to more efficient.

## Change source of calling label mismatch notifications of `MatchCallArgumentListener`

Their source is `CallArgumentMatchResult`.
It does same as original code.

I think using `RelabelingMapping` is better.
It is consistent with following stages.

Doing it reduce `matchCallArgument` codes and `CallArgumentMatchResult`'s fields.

### CODE REVIEW

**Comment by omochi** [comment]
I've been thinking about the design around here for the last few days.
And just recently, I realized that I might have made a fatal mistake.

Currently two label mismatch handlings are implemented.

(1) One is the `(incorrect | missing | extraneous) Label` of the` ArgumentFailureTracker`.
This implementation is empty.

(2) Another is `diagnoseArgumentLabelError`.
It compares each label inside it and may diagnose `diag :: missing_argument_labels`.

The two seem similar but have significant differences.
That is the behavior when the label order is wrong.

(1) is based on label matching which contains typo correction.
(2) is based on the order in which the labels were written.

I initially thought (2) was the way to go.
Because A's implementation was empty.

But I found this commit.
https://github.com/apple/swift/commit/84cd99a3eb3b223fc99b5fbe6c5a392a1d9f895a

So (1) is a new direction.

@xedin said `avoid doing all the this work` in here.
https://forums.swift.org/t/question-about-overlapped-multiple-fixits-from-diagnostics/34488/2

I missed the meaning of the word **"all"**.

In conclusion, I withdraw this PR and will rethink direction.

**Comment by xedin** [comment]
I went though the code briefly today, going to leave a comment here once I have all my thoughts together to help you out. Thank you for helping with this!

File: lib/Sema/ConstraintSystem.h [hunk 1]
```diff
@@ -4643,6 +4643,69 @@ static inline bool computeTupleShuffle(TupleType *fromTuple,
 /// TinyPtrVector for unsigned values.
 using ParamBinding = SmallVector<unsigned, 1>;
 
+/// It holds all interested computation result of \c matchCallArguments
+struct CallArgumentMatchResult {
```
Comment (xedin) [inline lib/Sema/ConstraintSystem.h]: Note that we'd like to allocate as little as possible in a "happy" case, so I think the result of `matchCallArguments` should stay the same (array of arrays with positions). We should track all the information in `ArgumentFailureTracker` listener when it's relevant. Which brings us to the work performed issue, I think until solver is in "diagnostic" mode we shouldn't try to fix or track anything and do as little as possible.

It might make sense to just scrap current implementation and start from scratch having in mind all the things it needs to support regarding diagnostics. Maybe have it encapsulated into a struct/class so it could be re-used. 

Another  issue is `diagnoseArgumentLabelError` - I think that shouldn't exist because it's too late to diagnose that post-factum when solution is already applied. Currently I see it being used only in `CSDiagnostics` and `TypeCheckAvailability` we need to take a closer look at the second use-case and see whether it even makes sense or not. All of the logic currently associated with `diagnoseArgumentLabelError` should be a part of `matchCallArguments`.

Comment (omochi) [reply lib/Sema/ConstraintSystem.h]: Thank you for your comment.

I understood that keeping memory low is important in "happy" case.

I finished reworking from scratch just now.
It doesn't use `diagnoseArgumentLabelError` already.
